Parallel Packet Processing

TimonPost opened the pull request:
Removed any dependencies to specs by abstracting away all socket logic to the `Server` type. 

The current network library does send and receiving in one thread. This could cause the program to not receive some packets at high server load when the thread is in the send loop. I moved the receiving logic and sending logic into separate threads so that we can receive and send in parallel.

_A diagram of the implemented code:_
![untitled diagram](https://user-images.githubusercontent.com/19969910/50450660-24887880-08e4-11e9-9509-4ec8697db561.png)

jojolepro commented on amethyst_network/src/network_socket.rs:
```diff
@@ -71,88 +45,95 @@ where
-                            _ => {
-                                error!("Could not receive datagram: {:?}", e);
+                            ServerSocketEvent::Error(error) => {
+                                error!("[ERROR NETWORKING SYSTEM] {:?}", error)
```
error! already shows from which file it is called in the logs

jojolepro commented on amethyst_network/src/net_event.rs:
```diff
@@ -45,10 +45,10 @@ pub enum NetEvent<T> {
         }
     }
-}
+}
```
Missing new line after this one

jojolepro commented on examples/net_client/main.rs:
```diff
@@ -72,32 +75,3 @@ impl<'a> System<'a> for SpamSystem {
-
-/// A simple system reading received events.
-/// Used to see events sent by the net_echo_server example.
-struct ReaderSystem {
```
We can't read received packets from `System`s now?

TimonPost replied:
Wasn't used so I figured to remove it. The net_server does not send packets back.

TimonPost replied:
I added it again

TimonPost commented:
@jojolepro yeah, we had some discussions about it. We might move away this code to its own crate some time. For now it could be placed in this crate.

LucioFranco commented on amethyst_network/src/server/server.rs:
```diff
@@ -0,0 +1,105 @@
+    /// 3: Set up some `channels` to communicate with underlying threads.
+    ///
+    /// The method uses the config provided when creating a `Server` instance.
+    pub fn run(&mut self) {
```
from what I noticed above we could probably just merge run with new? This seems like an OOP style which is not really needed.

LucioFranco commented on amethyst_network/src/server/server.rs:
```diff
@@ -0,0 +1,105 @@
+    }
+
+    /// Get the handle to the internals of the UDP-receiving threat.
+    pub fn get_udp_receive_handle(&self) -> &Option<ReceiveHandler> {
```
I would drop the `get_` and just have it `udp_receive_handle` since the get part can be infered from the return type not being a `&mut`.

LucioFranco commented on amethyst_network/src/server/server.rs:
```diff
@@ -0,0 +1,105 @@
+    }
+
+    /// Get the handle to the internals of the UDP-sending thread.
+    pub fn get_udp_send_handle(&self) -> &Option<SendHandler> {
```
same here.

LucioFranco commented on amethyst_network/src/server/server.rs:
```diff
@@ -0,0 +1,105 @@
+
+    /// Schedule a TCP-packet for sending.
+    pub fn send_tcp(&mut self, payload: &[u8]) -> NetworkResult<()> {
+        unimplemented!()
```
we should try and avoid putting these into master so I think if this PR isn't going to include this we should have an issue and remove this function. aka I don't want these to slip out into an actual release.

LucioFranco commented on amethyst_network/src/server/server.rs:
```diff
@@ -0,0 +1,105 @@
+                error!("Trying to send UDP while it is not enabled");
+            }
+            Some(ref send_handle) => {
+                send_handle.send(ServerSocketEvent::Packet(packet)).unwrap();
```
Can we use `?` instead of unwrap?

TimonPost replied:
Read my PR description 😉

LucioFranco replied:
should of put that at the top!!! hahaha totally missed that with the diagram 🤦‍♂️

LucioFranco commented on amethyst_network/src/server/server.rs:
```diff
@@ -0,0 +1,105 @@
+    fn shutdown(&mut self) {
+        // 1. shutdown all connections udp/tcp
+        // 2. clean up resources by closing thread's and such.
+        unimplemented!()
```
these two should also be removed and put into an issue.

LucioFranco commented on amethyst_network/src/server/server.rs:
```diff
@@ -0,0 +1,105 @@
+    }
+
+    /// Broadcast a TCP-packet to all connected clients.
+    fn broad_cast_tcp(&mut self, payload: &[u8], addr: SocketAddr) -> NetworkResult<()> {
```
I think broadcast is generally one word 😄

LucioFranco commented on amethyst_network/src/server/server_socket_event.rs:
```diff
@@ -0,0 +1,44 @@
+
+impl Clone for ServerSocketEvent {
+    fn clone(&self) -> Self {
+        ServerSocketEvent::Empty
```
This doesn't seem right, seems like a default impl but this is for clone?

TimonPost replied:
Clone is used in a vector allocation: `

 let mut buffer = vec![ServerSocketEvent::Empty; receiver.max_throughput];`

to this buffer, incoming events will be written.

LucioFranco replied:
I don't understand what you mean here? Clone should be to kinda copy the item not construct a new one?

fhaynes replied:
Clone should indeed copy the internal state of whatever is being cloned. I think this would be default, not clone.

LucioFranco commented on amethyst_network/src/server/server_socket_event.rs:
```diff
@@ -0,0 +1,44 @@
+    /// represents a client who is timing out
+    Timedout,
+    /// represents an default value for this enum.
+    None,
```
Can you explain in what situations you might get an event but its none? that seems redudant

TimonPost replied:
I was using that for events currently not implemented. But now I took a look at it I saw that it was only used for `QualityChange` event from laminar. I changed `None` to `QualityChange` now.

LucioFranco commented on amethyst_network/src/server/udp.rs:
```diff
@@ -0,0 +1,125 @@
+            udp_receiver.start_receiving(tx);
+        });
+
+        ReceiveHandler::new(5000, rx, thread_handle)
```
where does this 5000 come from?

TimonPost replied:
Don't know actually, and I'm not sure at how many packets we should break the receiving. This would be eventually done with congestion prevention.

TimonPost replied:
I will make it work via the `ServerConfig`

LucioFranco replied:
👍 we should add this to some sort of config.

LucioFranco commented on amethyst_network/src/server/udp.rs:
```diff
@@ -0,0 +1,125 @@
+            };
+
+            // iter over client events occurred from within laminar.
+            self.socket.events().into_iter().map(|e| {
```
nit but a for_each is better here since it describes that you are not mapping the values but actually just using the values.

LucioFranco commented:
about moving this to another crate, I think for the moment its fine, but I imagine we could provide some abstractions on top of laminar for some of this. But I think amethyst is a good playground in a sense to find what works.

TimonPost commented on amethyst_network/src/net_event.rs:
```diff
@@ -45,10 +45,10 @@ pub enum NetEvent<T> {
     pub fn custom(&self) -> Option<&T> {
-        if let NetEvent::Custom(ref t) = self {
-            Some(t)
+        if let &NetEvent::Custom(ref t) = self {
```
Not sure

TimonPost replied:
changed it back.

torkleyy replied:
Consider using `*self` instead (or you'll get a Clippy warning :wink:)

TimonPost commented on amethyst_network/src/server/server.rs:
```diff
@@ -0,0 +1,105 @@
+use std::net::SocketAddr;
+use std::sync::Arc;
+
+/// 'Server' abstracts TCP and UDP sockets away.
```
Yes I like that

TimonPost commented:
According to all unimplemented macro's I'll make sure all of them are getting issues bound to it. I forgot to mention it in the PR description.

LucioFranco commented:
@TimonPost I think it would be fine to group them into one big tracking issue with checkboxes etc.

TimonPost commented on amethyst_network/src/bundle.rs:
```diff
@@ -7,32 +7,56 @@ use amethyst_core::{
 pub struct NetworkBundle<T> {
     /// Local socket address
-    addr: SocketAddr,
+    pub config: ServerConfig,
```
can be private actually

TimonPost commented on amethyst_network/src/connection.rs:
```diff
@@ -32,12 +34,13 @@ pub struct NetConnection<E: 'static> {
 impl<E: Send + Sync + 'static> NetConnection<E> {
     /// Construct a new NetConnection. `SocketAddr` is the address that will be connected to.
-    pub fn new(target: SocketAddr) -> Self {
+    pub fn new(target_receiver: SocketAddr, target_sender: SocketAddr) -> Self {
```
I leave it since the connection created heer is one that simulates some connection to some endpoint. By having `target` it sounds clearer to me that you want to specify the target.

TimonPost commented on amethyst_network/src/lib.rs:
```diff
@@ -12,26 +12,27 @@ mod connection;
     filter::{FilterConnected, NetFilter},
     net_event::NetEvent,
-    network_socket::NetSocketSystem,
+    network_socket::*,
```
I had a reason for this when debugging some things, thanks for pointing out.

TimonPost commented on amethyst_network/src/lib.rs:
```diff
@@ -40,7 +41,10 @@ where
             let slice = s.as_slice();
             // send an unreliable `Packet` from laminar which is basically just a bare UDP packet.
-            match socket.send(&Packet::unreliable(*addr, slice.to_vec())) {
+            match sender.send(ServerSocketEvent::Packet(Packet::unreliable(
```
What do you mean by that?

LucioFranco replied:
Break out some of the parts to construct the type so 

```rust
let packet = Packet::unreliable(...);
let event = ServerSocketEvent::Packet(...);
match sender.send(event) {
...
}
```

TimonPost commented on amethyst_network/src/server/receive_handler.rs:
```diff
@@ -0,0 +1,55 @@
+/// Handler to access the internals of a receiving socket.
+pub struct ReceiveHandler {
+    /// handle that should be used for reading received packets from a given socket
+    pub receiver: Receiver<ServerSocketEvent>,
```
I marked them private because public it is not needed for now

TimonPost commented on amethyst_network/src/server/server.rs:
```diff
@@ -0,0 +1,105 @@
+pub struct Server {
+    config: ServerConfig,
+    // Handler to access the internals of the UDP receiving thread
+    udp_receiver: Option<ReceiveHandler>,
```
Removed the Option because Udp is always used. Tcp is still an option

LucioFranco commented on amethyst_network/src/server/send_handler.rs:
```diff
@@ -0,0 +1,32 @@
+    }
+}
+
+unsafe impl Sync for SendHandler {}
```
Let's take a look at it tomorrow together if you'd like.

fhaynes replied:
I don't think we should need unsafe for any of this. Let's please try to get rid of this.

LucioFranco commented on amethyst_network/src/error.rs:
```diff
@@ -0,0 +1,85 @@
+    ChannelSendError(mpsc::SendError<ServerSocketEvent>),
+}
+
+impl StdError for NetworkErrorKind {
```
honestly, I here would do `std::error::Error` I think that is better than renaming.

TimonPost replied:
Oke

LucioFranco commented on amethyst_network/src/error.rs:
```diff
@@ -0,0 +1,85 @@
+use crate::server::ServerSocketEvent;
+
+/// The `amethyst_network` result type.
+pub type NetworkError<T> = Result<T, NetworkErrorKind>;
```
This should be ` pub type Result<T> = Result<T, Error>;`

We should follow how `std::io::Error` does it. https://doc.rust-lang.org/src/std/io/error.rs.html#66-68

So just create:

```rust
pub struct Error {
  kind: ErrorKind,
}
```

and rename `NetworkErrorKind` to `ErrorKind`, since we have modules we can namespace via that and they can be renamed downstream as needed. This is how std does its errors.

LucioFranco commented:
@udoprog @torkleyy If either of you have time to checkout https://github.com/amethyst/amethyst/pull/1289/commits/40fa2f29ab6ae5a2d36af2cee5be22358b2129a6 related to error handling.

TimonPost commented on amethyst_network/src/error.rs:
```diff
@@ -0,0 +1,85 @@
+    }
+}
+
+impl From<Box<bincode::ErrorKind>> for NetworkErrorKind {
```
No this could not be done. `bincode::deserialize` is returning an error of type: `Result<T, Box<ErrorKind>`. I need to have a conversion from that to my `NetworkErrorKind`.

LucioFranco commented on amethyst_network/src/error.rs:
```diff
@@ -0,0 +1,85 @@
+impl StdError for NetworkErrorKind {
+    fn description(&self) -> &str {
+        match *self {
+            NetworkErrorKind::UdpError(_) => "An UDP error occurred",
```
though the error should still implement display so you can use that for now.

torkleyy commented on amethyst_network/src/error.rs:
```diff
@@ -0,0 +1,85 @@
+
+impl std::error::Error for ErrorKind {
+    fn description(&self) -> &str {
+        use std::error::Error;
```
`description` is deprecated, so I would remove this.

torkleyy commented on amethyst_network/src/error.rs:
```diff
@@ -0,0 +1,85 @@
+        }
+    }
+
+    fn cause(&self) -> Option<&dyn std::error::Error> {
```
`cause` is deprecated; `source` is used instead. This implementation will create a `source` that returns `None`.

torkleyy commented on amethyst_network/src/error.rs:
```diff
@@ -0,0 +1,85 @@
+    }
+}
+
+impl Display for ErrorKind {
```
I'd recommend using `err-derive` for deriving this and the `source` implementation.

torkleyy commented on amethyst_network/src/error.rs:
```diff
@@ -0,0 +1,85 @@
+use crate::server::ServerSocketEvent;
+
+/// The `amethyst_network` result type.
+pub type Error<T> = Result<T, ErrorKind>;
```
This is confusing; please call it `Result`.

torkleyy replied:
Consequently, `ErrorKind` can be called `Error`. `ErrorKind` is usually used for tags-only.

TimonPost replied:
Thanks for pointing out it should be indeed Result, must have accidentally renamed it to Error

torkleyy commented on amethyst_network/src/error.rs:
```diff
@@ -0,0 +1,85 @@
+    fn fmt(&self, fmt: &mut Formatter<'_>) -> fmt::Result {
+        match *self {
+            ErrorKind::UdpError(ref error) => write!(fmt, "UDP-error occurred {}", error),
+            ErrorKind::IoError(ref error) => write!(fmt, "IO-error occurred {}", error),
```
One last thing, printing an error that's also the `source()` is discouraged since that will duplicate all errors when printing the chain.

udoprog commented on amethyst_network/src/error.rs:
```diff
@@ -0,0 +1,85 @@
+    /// Error that could occur whit IO.
+    IoError(io::Error),
+    /// Error that could occur when serializing whit `bincode`
+    SerializeError(Box<bincode::ErrorKind>),
```
This has an alias, `bincode::Error`, that's probably what we'd want to use.

TimonPost replied:
I will take a look at it, because deserialize from bincode returns a boxed error kind. Maybe I can try do dereference it.

udoprog replied:
No need to deref, using the alias directly works (assuming it's available in our version of bincode):
https://docs.rs/bincode/1.0.1/bincode/type.Error.html

udoprog commented on amethyst_network/src/error.rs:
```diff
@@ -0,0 +1,85 @@
+
+/// Wrapper for all errors who could occur in `amethyst_network`.
+#[derive(Debug)]
+pub enum ErrorKind {
```
Make use of the [`__NonExhaustive` pattern](https://github.com/rust-lang/rfcs/blob/master/text/2008-non-exhaustive.md#how-we-do-this-today) so we guard against compatibility issues in case someone pattern matches over the type.

TimonPost commented:
Any other people who want to review those changes, I like this PR to merged.

udoprog commented on amethyst_network/src/error.rs:
```diff
@@ -0,0 +1,78 @@
+impl std::error::Error for ErrorKind {
+    fn source(&self) -> Option<&(dyn std::error::Error + 'static)> {
+        Some(self)
+    }
```
Right, that means that you should remove `cause` completely and _only_ implement `source`.

The default implementation of `source` is returning `None`: https://doc.rust-lang.org/src/std/error.rs.html#206
The default implementation of `cause` calls `source`: https://doc.rust-lang.org/src/std/error.rs.html#143

udoprog replied:
This is not fixed yet afaics?

TimonPost replied:
😑 I'm totally confused now, I'll take another try. I think I am missing interpreting  your suggestions. Do you have an code exaple?

TimonPost replied:
And is this really needed to change, since when the new error handling crate from your PR is merged all this error handling code will probably change.

udoprog replied:
This is what it should look like (it should look familiar since it's what you initially did in `cause`):
```rust
    fn source(&self) -> Option<&(dyn std::error::Error + 'static)> {
        match *self {
            ErrorKind::UdpError(ref e) => Some(e),
            ErrorKind::IoError(ref e) => Some(e),
            ErrorKind::ChannelSendError(ref e) => Some(e),
            ErrorKind::SerializeError(ref e) => Some(e),
            _ => None,
        }
    }
```

This is necessary because you are implementing a custom error type. Not using the dynamic one introduced in #1220.

The reason why your approach would cause an infinite loop is because you walk through the error chain like this:

```rust
let mut e = ...;

println!("Error: {}", e);

while let Some(s) = e.source() {
    println!("Caused by: {}", s);
    e = s;
}
```

And if `source` returns `Some(self)` this is gonna loop forever.

Check out the documentation for source for a more complete example on this:
https://doc.rust-lang.org/std/error/trait.Error.html#method.source

TimonPost replied:
Thanks for writing that up!!. I'll see what I can do whit this.

fhaynes commented on amethyst_network/src/lib.rs:
```diff
@@ -50,9 +56,9 @@ where
 
 /// Attempts to deserialize an event from the raw byte data.
-fn deserialize_event<T>(data: &[u8]) -> Result<NetEvent<T>, Box<ErrorKind>>
+fn deserialize_event<T>(data: &[u8]) -> Result<NetEvent<T>>
```
Non-public function so can use `//` not `///`

udoprog replied:
Should probably still prefer `///` since they are contextual to the function and used by e.g. RLS. Don't know if that works as intended with `//`.

fhaynes commented on amethyst_network/src/network_socket.rs:
```diff
@@ -71,88 +46,88 @@ where
+                for event in buffer[0..received].iter() {
+                    match event {
+                        ServerSocketEvent::Packet(packet) => {
+                            if let Err(_) = receive_queue.send(packet.clone()) {
```
Can we not throw the error away and print it? Would be helpful in debugging.

fhaynes commented on amethyst_network/src/network_socket.rs:
```diff
@@ -162,50 +137,60 @@ where
             }
+
+            // this will prevent our system to be stuck in the iterator.
+            // After 10000 packets we will continue and leaf the other packets for the next run.
```
Should be `leave` not `leaf`

fhaynes commented on amethyst_network/src/network_socket.rs:
```diff
@@ -162,50 +137,60 @@ where
+            // this will prevent our system to be stuck in the iterator.
+            // After 10000 packets we will continue and leaf the other packets for the next run.
+            // eventually some congestion prevention should be done.
+            if counter >= 10000 {
```
Can we make this a constant somewhere? Or even better, something we can set via the config?

fhaynes commented on amethyst_network/src/network_socket.rs:
```diff
@@ -162,50 +137,60 @@ where
                     // Get the event
-                    let net_event = deserialize_event::<E>(raw_event.data.as_slice());
+                    let net_event = deserialize_event::<E>(raw_event.payload());
                     match net_event {
```
Just to make sure I understand, the idea here is that you add packets until you can deserialize an Event?

TimonPost replied:
The idea here is that your packet will be deserialized into a `NetEvent<E>` from the received packet payload.

fhaynes commented on amethyst_network/src/server/config.rs:
```diff
@@ -0,0 +1,26 @@
+impl Default for ServerConfig {
+    fn default() -> Self {
+        ServerConfig {
+            // py passing in :0 port the OS will give an available port.
```
Should be 'By' not 'py'

fhaynes commented on amethyst_network/src/server/host.rs:
```diff
@@ -0,0 +1,100 @@
+/// This module is about the communication whit this host to some other endpoints.
```
should be `with` not `whit`

fhaynes commented on amethyst_network/src/server/host.rs:
```diff
@@ -0,0 +1,100 @@
+    /// The method uses the config provided when creating a `host` instance.
+    pub fn run(config: ServerConfig) -> Result<Host> {
+        if config.enable_tcp {
+            // todo: create TCP support.
```
Can we print out a message here so the user doesn't think TCP is working when it really isn't?

fhaynes commented on amethyst_network/src/server/host.rs:
```diff
@@ -0,0 +1,100 @@
+        Ok(())
+    }
+
+    /// Shutdown the server and destroy all related resources.
```
Private function, can do `//`

fhaynes commented on amethyst_network/src/server/host.rs:
```diff
@@ -0,0 +1,100 @@
+    }
+
+    /// Broadcast a TCP-packet to all connected clients.
+    fn broadcast_tcp(&mut self, payload: &[u8], addr: SocketAddr) -> Result<()> {
```
Private function, can do `//`

fhaynes commented on amethyst_network/src/server/host.rs:
```diff
@@ -0,0 +1,100 @@
+        unimplemented!()
+    }
+
+    /// Broadcast a UDP-packet to all connected clients.
```
Private function, can do `//`

fhaynes commented on amethyst_network/src/server/host.rs:
```diff
@@ -0,0 +1,100 @@
+    }
+
+    /// Broadcast a UDP-packet to all connected clients.
+    fn broadcast_cast_upd(
```
should be `udp` not `upd`

fhaynes commented on amethyst_network/src/server/mod.rs:
```diff
@@ -0,0 +1,25 @@
+pub use self::server_socket_event::{ClientEvent, ServerSocketEvent};
+use std::sync::mpsc::{Receiver, Sender};
+
+pub trait PacketReceiving {
```
Comment the trait please since it is public

fhaynes commented on amethyst_network/src/server/mod.rs:
```diff
@@ -0,0 +1,25 @@
+    fn start_receiving(&mut self, tx: Sender<ServerSocketEvent>);
+}
+
+pub trait PacketSending {
```
Comment this as it is a public trait please

fhaynes commented on amethyst_network/src/server/udp.rs:
```diff
@@ -0,0 +1,128 @@
+            udp_receiver.start_receiving(tx);
+        });
+
+        Ok(ReceiveHandler::new(5000, rx, thread_handle))
```
This should be a constant or configurable, I think

fhaynes commented on amethyst_network/src/server/udp.rs:
```diff
@@ -0,0 +1,128 @@
+                            error!("Something went wrong when trying to send a packet with UDP socket. Reason: {:?}", e)
+                        }
+                    }
+                    _ => {}
```
Might be worth printing an `error!` here, just in case

LucioFranco commented on amethyst_network/src/bundle.rs:
```diff
@@ -7,32 +7,58 @@ use amethyst_core::{
-        NetworkBundle { addr, filters }
+    /// Creates a new NetworkBundle.
+    ///
+    /// `receive_addr:` this is the address on which we will receive incoming packets.
```
nit lets move the `:` after the ticks 😄

LucioFranco commented on amethyst_network/src/lib.rs:
```diff
@@ -49,10 +55,10 @@ where
     T: DeserializeOwned,
 {
-    deserialize::<NetEvent<T>>(data)
+    Ok(deserialize::<NetEvent<T>>(data)?)
```
This should be `deserialize::<NetEvent<T>>(data).map_err(Error::from)`

LucioFranco commented on amethyst_network/src/network_socket.rs:
```diff
@@ -70,89 +45,90 @@ where
-                "Unable to set `UdpSocket` to non-blocking mode",
-            )
-        })?;
+        let server_sender = NetSocketSystem::<E>::start_sending(&server);
```
this block isn't clear to me why do you send a reference in one call but not in the other? Does a clone happen on the other side if so I would like this to be moved up to this piece of code.

TimonPost replied:
`start_sending` takes the reference and `start_receiving` will eventually own the instance. It is just a way to play whit the borrowing.

LucioFranco replied:
Yeah, I see now this confuses me, to be honest, I find it unclear what is being borrowed and for how long. Since you are passing a channel handle down do you think that this function could take that handle instead?

TimonPost replied:
Agree, I have to experiment with this further, I tried some things a few seconds ago but it is a little difficult to borrow those Handles.

TimonPost replied:
done, I had to move `SendHandler` into an Arc. Although for a `Receiver` it is more difficult to move it around threads since it does not implement `Sync`. For this, I had to move it into a `Mutex`

LucioFranco commented on amethyst_network/src/network_socket.rs:
```diff
@@ -70,89 +45,90 @@ where
-                            }
+    /// Start a thread to send all queued packets.
+    fn start_sending(server: &Host) -> Sender<InternalSocketEvent<E>> {
+        let (tx, send_queue) = channel();
```
small nit but can you use `mpsc::channel()` here?

LucioFranco commented on amethyst_network/src/network_socket.rs:
```diff
@@ -70,89 +45,90 @@ where
+        thread::spawn(move || {
+            let receiver = server.udp_receive_handle();
+
+            let mut buffer = vec![ServerSocketEvent::Empty; receiver.max_throughput];
```
Nit: I'd like to see this renamed to buffer_size not needed for this PR but just a note.

LucioFranco commented on amethyst_network/src/network_socket.rs:
```diff
@@ -70,89 +45,90 @@ where
+    }
+
+    /// Starts a thread which receives incoming packets and sends them onto the 'Receiver' channel.
+    fn start_receiving(server: Host) -> Receiver<Packet> {
```
I think this should receive the udp receive handle instead of the host, as that is a clearer interface.

LucioFranco commented on amethyst_network/src/network_socket.rs:
```diff
@@ -162,50 +138,60 @@ where
+                if net_connection.target_sender == raw_event.addr() {
                     // Get the event
-                    let net_event = deserialize_event::<E>(raw_event.data.as_slice());
+                    let net_event = deserialize_event::<E>(raw_event.payload());
```
not important but these can be merged.

TimonPost replied:
could you elaborate on what you mean whit merge?

LucioFranco replied:
```rust
match deserialize_event::<E>(raw_event.payload()) { ... }
```

LucioFranco commented on amethyst_network/src/server/send_handler.rs:
```diff
@@ -0,0 +1,41 @@
+
+    /// Get the sending channel of this `SendHandler` to which you can send  
+    pub fn get_sender(&self) -> &SyncSender<ServerSocketEvent> {
+        return &self.sender;
```
drop this return

LucioFranco commented on amethyst_network/src/server/send_handler.rs:
```diff
@@ -0,0 +1,41 @@
+
+    /// Send an event on the internal channel, by doing this it will be scheduled for sending.
+    pub fn send(&self, event: ServerSocketEvent) -> Result<()> {
+        self.sender.send(event)?;
```
these two lines can just be `self.sender.send(event)`

LucioFranco commented on amethyst_network/src/server/udp.rs:
```diff
@@ -0,0 +1,136 @@
+pub struct UdpReceiver {
+    // socket used for receiving packets
+    socket: UdpSocket,
+    pub max_troughput: u16,
```
This either needs a comment or if this can not change during the life of a UdpReceiver then I would move this to a `max_throughput()` function.

TimonPost commented on amethyst_network/src/error.rs:
```diff
@@ -0,0 +1,75 @@
+    }
+}
+
+impl From<laminar::error::NetworkError> for ErrorKind {
```
This should be an issue for laminar, but I agree laminar should have also have `ErrorKind`, `Result`

TimonPost commented on amethyst_network/src/error.rs:
```diff
@@ -0,0 +1,75 @@
+    }
+}
+
+impl From<Box<bincode::ErrorKind>> for ErrorKind {
```
Can't be done to my knowledge. deserialize returns a boxed error and the compiler is searching for `From<Box<ErrorKind>>` implementations to translate this error into our own ErrorKind.

TimonPost commented:
@LucioFranco I removed the `get_events` method with an iterator call, this is IMO much better. Also, I refactored the functions `start_sending` and `start_receiving` to not take the `Host` but instead just the `ReceiveHandler` or `SendHandler` as requested. However, I had to move the receiver into an `Arc<Mutex<Receiver>>` because `Receiver` does not implement `Sync`.

LucioFranco commented on amethyst_network/src/bundle.rs:
```diff
@@ -7,30 +7,50 @@ use amethyst_core::{
 };
 
-use crate::{filter::NetFilter, NetSocketSystem};
+use super::server::ServerConfig;
```
super super minor but I'd like these to be merged if possible.

LucioFranco commented on amethyst_network/src/network_socket.rs:
```diff
@@ -69,89 +50,91 @@ where
+        let (receive_queue, rx) = mpsc::channel();
+
+        thread::spawn(move || {
+            if let Ok(receiver) = receiver.lock() {
```
this lock will never be unlocked unless there is a send error, this seems a bit odd? Though I think this function will only get called once? So I think we should either document that or panic if it gets called twice.

TimonPost replied:
Indeed, this will not be possible. 

The whole point why the `Receiver` is not sync is that there could not be different receivers. This is the case here. We have one receiving thread, and so requiring the lock for the receiver there is a valid use case. If someone else tries to receive it is not possible. And if he does he should not be able.

LucioFranco replied:
Yeah what I’m saying is that we should be explicit about this behavior so there should be a comment about it and a panic if it gets called more than once.

TimonPost replied:
Yes, I'll add that

LucioFranco commented on amethyst_network/src/network_socket.rs:
```diff
@@ -161,50 +144,56 @@ where
         }
 
-        for raw_event in self.rx.try_iter() {
+        for (counter, raw_event) in self.transport_receiver.try_iter().enumerate() {
```
👍 good solution I didn't think of it this way and this is def simpler! Good stuff!

TimonPost commented:
bors r= LucioFranco, fhaynes

TimonPost commented:
bors r= LucioFranco, fhaynes

TimonPost commented:
bors r= LucioFranco, fhaynes

TimonPost commented:
I hope I don't make bors angry today. I put him at work too much haha.

TimonPost commented:
bors r= LucioFranco, fhaynes
